File: projects/2602/09-survivor-game/src/core/camera.py
"""Camera system that follows the player with smooth interpolation."""

import pygame

from src.core.settings import (
    SCREEN_WIDTH, SCREEN_HEIGHT,
    WORLD_WIDTH, WORLD_HEIGHT,
    CAMERA_LERP,
)


class Camera:
    """Camera that smoothly follows a target within world bounds."""

    def __init__(self):
        # Camera position represents the top-left corner of the viewport
        self.x = 0.0
        self.y = 0.0
        self.width = SCREEN_WIDTH
        self.height = SCREEN_HEIGHT

    def update(self, target_x: float, target_y: float, dt: float):
        """Smoothly follow the target position (usually player center).

        Args:
            target_x: Target world x position to center on.
            target_y: Target world y position to center on.
            dt: Delta time in seconds.
        """
        # Desired camera position: center the target on screen
        desired_x = target_x - self.width / 2
        desired_y = target_y - self.height / 2

        # Smooth interpolation (framerate-independent lerp)
        # Using 1 - (1 - CAMERA_LERP)^(dt*60) for consistent feel across framerates
        factor = 1.0 - (1.0 - CAMERA_LERP) ** (dt * 60)
        self.x += (desired_x - self.x) * factor
        self.y += (desired_y - self.y) * factor

        # Clamp to world boundaries
        self.x = max(0, min(self.x, WORLD_WIDTH - self.width))
        self.y = max(0, min(self.y, WORLD_HEIGHT - self.height))

    def apply(self, world_x: float, world_y: float) -> tuple[float, float]:
        """Convert world coordinates to screen coordinates.

        Args:
            world_x: X position in world space.
            world_y: Y position in world space.

        Returns:
            Tuple of (screen_x, screen_y).
        """
        return world_x - self.x, world_y - self.y

    def apply_rect(self, rect: pygame.Rect) -> pygame.Rect:
        """Convert a world-space rect to screen-space rect.

        Args:
            rect: A pygame.Rect in world coordinates.

        Returns:
            New pygame.Rect in screen coordinates.
        """
        return pygame.Rect(rect.x - self.x, rect.y - self.y, rect.width, rect.height)

    def get_visible_rect(self) -> pygame.Rect:
        """Return the world-space rectangle currently visible on screen."""
        return pygame.Rect(int(self.x), int(self.y), self.width, self.height)

    def is_visible(self, world_x: float, world_y: float, margin: float = 50) -> bool:
        """Check if a world position is visible on screen (with margin).

        Args:
            world_x: X position in world space.
            world_y: Y position in world space.
            margin: Extra pixels beyond screen edge to still count as visible.

        Returns:
            True if the position is within the visible area + margin.
        """
        return (self.x - margin <= world_x <= self.x + self.width + margin and
                self.y - margin <= world_y <= self.y + self.height + margin)

    def center_on(self, world_x: float, world_y: float):
        """Immediately center the camera on a world position (no lerp).

        Args:
            world_x: X position in world space.
            world_y: Y position in world space.
        """
        self.x = world_x - self.width / 2
        self.y = world_y - self.height / 2
        # Clamp to world boundaries
        self.x = max(0, min(self.x, WORLD_WIDTH - self.width))
        self.y = max(0, min(self.y, WORLD_HEIGHT - self.height))

File: projects/2602/10-starcraft-rts/js/core/camera.js
// === StarCraft RTS Clone - Camera ===
// Handles viewport position, edge scrolling, keyboard scrolling, zoom,
// and world<->screen coordinate conversions.
// Exposes at window.SC.Camera

window.SC = window.SC || {};

window.SC.Camera = {
  // --- Public state ---
  x: 0,           // top-left world X of viewport
  y: 0,           // top-left world Y of viewport
  zoom: 1.0,      // current zoom level
  width: 0,       // viewport width in pixels (canvas width)
  height: 0,      // viewport height in pixels (canvas height)

  /**
   * Initialize the camera with viewport dimensions.
   * @param {number} canvasWidth
   * @param {number} canvasHeight
   */
  init(canvasWidth, canvasHeight) {
    this.width = canvasWidth;
    this.height = canvasHeight;
    this.x = 0;
    this.y = 0;
    this.zoom = 1.0;
  },

  /**
   * Convert world coordinates to screen (canvas) coordinates.
   * @param {number} worldX
   * @param {number} worldY
   * @returns {{ x: number, y: number }}
   */
  worldToScreen(worldX, worldY) {
    return {
      x: (worldX - this.x) * this.zoom,
      y: (worldY - this.y) * this.zoom,
    };
  },

  /**
   * Convert screen (canvas) coordinates to world coordinates.
   * @param {number} screenX
   * @param {number} screenY
   * @returns {{ x: number, y: number }}
   */
  screenToWorld(screenX, screenY) {
    return {
      x: screenX / this.zoom + this.x,
      y: screenY / this.zoom + this.y,
    };
  },

  /**
   * Get the visible rectangle in world coordinates.
   * @returns {{ x: number, y: number, w: number, h: number }}
   */
  getVisibleRect() {
    return {
      x: this.x,
      y: this.y,
      w: this.width / this.zoom,
      h: this.height / this.zoom,
    };
  },

  /**
   * Update camera position from edge scroll, keyboard, and clamping.
   * Called once per frame.
   * @param {number} dt - delta time in seconds
   */
  update(dt) {
    const C = SC.CONST;
    let dx = 0;
    let dy = 0;

    // --- Edge scrolling ---
    if (SC.Input) {
      const mx = SC.Input.mouse.x;
      const my = SC.Input.mouse.y;
      const margin = C.CAMERA_EDGE_SCROLL_MARGIN;
      const speed = C.CAMERA_EDGE_SCROLL_SPEED;

      if (mx <= margin) {
        dx -= speed * dt;
      } else if (mx >= this.width - margin) {
        dx += speed * dt;
      }

      if (my <= margin) {
        dy -= speed * dt;
      } else if (my >= this.height - margin) {
        dy += speed * dt;
      }
    }

    // --- Keyboard scrolling ---
    if (SC.Input) {
      const kSpeed = C.CAMERA_KEY_SCROLL_SPEED;
      if (SC.Input.isKeyDown('ArrowLeft') || SC.Input.isKeyDown('a')) {
        // 'a' is attack-move hotkey, so only use arrows
      }
      if (SC.Input.isKeyDown('ArrowLeft')) {
        dx -= kSpeed * dt;
      }
      if (SC.Input.isKeyDown('ArrowRight')) {
        dx += kSpeed * dt;
      }
      if (SC.Input.isKeyDown('ArrowUp')) {
        dy -= kSpeed * dt;
      }
      if (SC.Input.isKeyDown('ArrowDown')) {
        dy += kSpeed * dt;
      }
    }

    // Apply movement
    this.x += dx;
    this.y += dy;

    // --- Clamp to world bounds ---
    this._clamp();

    // Emit event
    if (dx !== 0 || dy !== 0) {
      if (SC.EventBus) {
        SC.EventBus.emit(SC.CONST.EVENT.CAMERA_MOVED, { x: this.x, y: this.y });
      }
    }
  },

  /**
   * Center the camera on a world position.
   * @param {number} worldX
   * @param {number} worldY
   */
  centerOn(worldX, worldY) {
    this.x = worldX - (this.width / this.zoom) / 2;
    this.y = worldY - (this.height / this.zoom) / 2;
    this._clamp();

    if (SC.EventBus) {
      SC.EventBus.emit(SC.CONST.EVENT.CAMERA_MOVED, { x: this.x, y: this.y });
    }
  },

  /**
   * Check if a world point is within the visible viewport.
   * @param {number} worldX
   * @param {number} worldY
   * @param {number} [margin=0] - extra margin in world pixels
   * @returns {boolean}
   */
  isVisible(worldX, worldY, margin) {
    margin = margin || 0;
    const rect = this.getVisibleRect();
    return (
      worldX >= rect.x - margin &&
      worldX <= rect.x + rect.w + margin &&
      worldY >= rect.y - margin &&
      worldY <= rect.y + rect.h + margin
    );
  },

  /**
   * Apply zoom change. Zoom toward/away from the mouse position.
   * @param {number} delta - +1 to zoom in, -1 to zoom out
   * @param {number} screenX - mouse screen X for zoom anchor
   * @param {number} screenY - mouse screen Y for zoom anchor
   */
  applyZoom(delta, screenX, screenY) {
    const C = SC.CONST;

    // World point under mouse before zoom
    const beforeWorld = this.screenToWorld(screenX, screenY);

    // Change zoom
    const oldZoom = this.zoom;
    this.zoom += delta * C.CAMERA_ZOOM_STEP;
    this.zoom = Math.max(C.CAMERA_ZOOM_MIN, Math.min(C.CAMERA_ZOOM_MAX, this.zoom));

    if (this.zoom === oldZoom) return;

    // Adjust camera so the world point under mouse stays in same screen position
    this.x = beforeWorld.x - screenX / this.zoom;
    this.y = beforeWorld.y - screenY / this.zoom;

    this._clamp();
  },

  /**
   * Internal: clamp camera position to world bounds.
   */
  _clamp() {
    const C = SC.CONST;
    const viewW = this.width / this.zoom;
    const viewH = this.height / this.zoom;

    // Don't allow camera to go past world edges
    this.x = Math.max(0, Math.min(C.MAP_WIDTH_PX - viewW, this.x));
    this.y = Math.max(0, Math.min(C.MAP_HEIGHT_PX - viewH, this.y));
  },
};
